File: packages/engine/src/proposePlan.ts
import type OpenAI from "openai";
import { ParsePlanSchema, validateParsePlan, type ExecReport, type ParsePlan } from "@runoff/core";
import type { SheetGrid } from "./parsePlan.js";
import { rawCell } from "./parsePlan.js";
import { MODEL } from "./prompts.js";

const SAMPLE_ROWS = 30;
const SAMPLE_COLS = 15;
const SAMPLE_CAP = 6000;
const CELL_CAP = 20;

export function buildGridSample(grids: SheetGrid[], hints: string): string {
  const parts: string[] = [];
  for (const { sheet, grid } of grids) {
    const rows = Math.max(grid.length, 0);
    const cols = grid.reduce((m, r) => Math.max(m, (r ?? []).length), 0);
    const lines = [`## sheet: ${sheet} (${rows}×${cols})`];
    for (let r = 0; r < Math.min(rows, SAMPLE_ROWS); r++) {
      const cells = (grid[r] ?? []).slice(0, SAMPLE_COLS).map((v) => rawCell(v).slice(0, CELL_CAP));
      lines.push(`R${r + 1}: ${cells.join(" | ")}`);
    }
    parts.push(lines.join("\n"));
  }
  parts.push(`## detector hints\n${hints}`);
  return parts.join("\n\n").slice(0, SAMPLE_CAP);
}

const PLAN_CONTRACT =
  `Return JSON matching: {"version":1,"tables":[{"name":snake_case,"anchor":{"sheet"?:slug,` +
  `"headerSignature":[normalized header cell texts],"minMatch":int},"headerRows":1-3,` +
  `"exclude":[{"column":canonical|null,"pattern":regex}],"columns":[{"from":normalized merged header,` +
  `"name":snake_case,"type":"TEXT"|"INTEGER"|"REAL","parse"?:"number"|"currency"|"percent"|"date"}],` +
  `"unpivot"?:{"keep":[canonical],"valuePattern":regex,"keyColumn":snake_case,"valueColumn":snake_case,` +
  `"valueType":"INTEGER"|"REAL"|"TEXT","valueParse"?:"number"|"currency"|"percent"},` +
  `"periodColumn"?:canonical (its column MUST have parse "date"),"onPeriodMismatch"?:"keep"|"exclude"}]}. ` +
  `Header/from texts are lowercase, whitespace-collapsed. One entry per REAL data table; title rows, ` +
  `notes and total rows are NOT tables. Exclude total/subtotal rows with an exclude rule on a key column. ` +
  `Set minMatch ≈ two-thirds of the signature length. Unpivot wide period-like column layouts. ` +
  `Use parse for formatted values ("$1,234"→currency, "12%"→percent, dates→date).`;

/**
 * One structured call proposing (or amending) a ParsePlan. Zod + structural
 * validation; one retry on invalid; null on any failure — callers fall back to
 * the plan-less flow.
 */
export async function proposeParsePlan(opts: {
  client: OpenAI;
  filename: string;
  gridSample: string;
  existingPlan?: ParsePlan;
  fitDetail?: string[];
  execReport?: ExecReport;
  feedback?: string;
}): Promise<ParsePlan | null> {
  const amendment = opts.existingPlan
    ? ` You are AMENDING a working plan. You MUST keep every existing logical table name and canonical column name; ` +
      `re-anchor and re-map "from" texts onto them. You may ADD new columns; never remove or rename existing ones.`
    : "";
  const system =
    `You write parse plans that turn one uploaded spreadsheet/CSV into clean database tables. ` +
    `A deterministic engine executes your plan; anchors locate each table's header row by cell-text ` +
    `signature, so plans survive moved/renamed sheets. ${PLAN_CONTRACT}${amendment}`;
  const userParts = [`Filename: ${opts.filename}`, `File sample:\n${opts.gridSample}`];
  if (opts.existingPlan) userParts.push(`Existing plan:\n${JSON.stringify(opts.existingPlan)}`);
  if (opts.fitDetail?.length) userParts.push(`Fit problems:\n${opts.fitDetail.join("\n")}`);
  if (opts.execReport) userParts.push(`Execution report of the previous attempt:\n${JSON.stringify(opts.execReport)}`);
  if (opts.feedback) userParts.push(`User feedback:\n${opts.feedback}`);

  for (let attempt = 0; attempt < 2; attempt++) {
    let raw: string;
    try {
      const res = await (opts.client as any).chat.completions.create({
        model: MODEL,
        messages: [
          { role: "system", content: system },
          { role: "user", content: userParts.join("\n\n") },
        ],
        response_format: { type: "json_object" },
        max_completion_tokens: 2000,
      });
      raw = res.choices?.[0]?.message?.content ?? "";
    } catch {
      return null;
    }
    try {
      const parsed = ParsePlanSchema.safeParse(JSON.parse(raw));
      if (!parsed.success) continue;
      validateParsePlan(parsed.data);
      return parsed.data;
    } catch {
      continue;
    }
  }
  return null;
}

/** Should the self-check round fire? See plan Global Constraints for the pinned criteria. */
export function isDegenerate(report: ExecReport): boolean {
  return report.tables.some(
    (t) =>
      t.problems.length > 0 ||
      t.rowsKept === 0 ||
      t.coercionFailures.some((f) => t.rowsKept > 0 && f.count >= t.rowsKept),
  );
}

File: backend/runoff_api/engine/propose_plan.py
"""Port of packages/engine/src/proposePlan.ts — render a grid sample, make one
structured (or amending) ParsePlan proposal call, and the degeneracy self-check.

``propose_parse_plan`` does one structured call proposing (or amending) a
ParsePlan; Zod (pydantic) + structural validation; one retry on invalid; None on
any failure — callers fall back to the plan-less flow. The amend/feedback variant
lives entirely in the optional ``existing_plan`` / ``fit_detail`` / ``exec_report``
/ ``feedback`` params, mirroring the TS opts.

Runtime shapes are plain dicts with camelCase keys (grids ``{"sheet","grid"}``,
ExecReport, ParsePlan). The request payload (model, messages, response_format,
max_completion_tokens) is field-for-field identical to the TS ``create(...)``
object literal — a later task byte-checks it.
"""

import json

from runoff_api.core.types.parse_plan import ParsePlan, validate_parse_plan
from runoff_api.engine.parse_plan import raw_cell
from runoff_api.engine.prompts import MODEL

SAMPLE_ROWS = 30
SAMPLE_COLS = 15
SAMPLE_CAP = 6000
CELL_CAP = 20


def build_grid_sample(grids: list[dict], hints: str) -> str:
    parts: list[str] = []
    for g in grids:
        sheet = g["sheet"]
        grid = g["grid"]
        rows = max(len(grid), 0)
        cols = 0
        for r in grid:
            cols = max(cols, len(r if r is not None else []))
        lines = [f"## sheet: {sheet} ({rows}×{cols})"]
        for r in range(min(rows, SAMPLE_ROWS)):
            row = grid[r] if grid[r] is not None else []
            cells = [raw_cell(v)[:CELL_CAP] for v in row[:SAMPLE_COLS]]
            lines.append(f"R{r + 1}: {' | '.join(cells)}")
        parts.append("\n".join(lines))
    parts.append(f"## detector hints\n{hints}")
    return "\n\n".join(parts)[:SAMPLE_CAP]


PLAN_CONTRACT = (
    'Return JSON matching: {"version":1,"tables":[{"name":snake_case,"anchor":{"sheet"?:slug,'
    '"headerSignature":[normalized header cell texts],"minMatch":int},"headerRows":1-3,'
    '"exclude":[{"column":canonical|null,"pattern":regex}],"columns":[{"from":normalized merged header,'
    '"name":snake_case,"type":"TEXT"|"INTEGER"|"REAL","parse"?:"number"|"currency"|"percent"|"date"}],'
    '"unpivot"?:{"keep":[canonical],"valuePattern":regex,"keyColumn":snake_case,"valueColumn":snake_case,'
    '"valueType":"INTEGER"|"REAL"|"TEXT","valueParse"?:"number"|"currency"|"percent"},'
    '"periodColumn"?:canonical (its column MUST have parse "date"),"onPeriodMismatch"?:"keep"|"exclude"}]}. '
    "Header/from texts are lowercase, whitespace-collapsed. One entry per REAL data table; title rows, "
    "notes and total rows are NOT tables. Exclude total/subtotal rows with an exclude rule on a key column. "
    "Set minMatch ≈ two-thirds of the signature length. Unpivot wide period-like column layouts. "
    'Use parse for formatted values ("$1,234"→currency, "12%"→percent, dates→date). '
    "Patterns are ECMAScript regex — no PCRE inline flags like (?i); matching is already case-insensitive."
)


def propose_parse_plan(
    *,
    client,
    filename: str,
    grid_sample: str,
    existing_plan: dict | None = None,
    fit_detail: list[str] | None = None,
    exec_report: dict | None = None,
    feedback: str | None = None,
) -> dict | None:
    amendment = (
        " You are AMENDING a working plan. You MUST keep every existing logical table name and canonical "
        "column name; "
        're-anchor and re-map "from" texts onto them. You may ADD new columns; never remove or rename '
        "existing ones."
        if existing_plan
        else ""
    )
    system = (
        "You write parse plans that turn one uploaded spreadsheet/CSV into clean database tables. "
        "A deterministic engine executes your plan; anchors locate each table's header row by cell-text "
        f"signature, so plans survive moved/renamed sheets. {PLAN_CONTRACT}{amendment}"
    )
    user_parts = [f"Filename: {filename}", f"File sample:\n{grid_sample}"]
    if existing_plan:
        plan_json = json.dumps(existing_plan, separators=(",", ":"), ensure_ascii=False)
        user_parts.append(f"Existing plan:\n{plan_json}")
    if fit_detail:
        user_parts.append("Fit problems:\n" + "\n".join(fit_detail))
    if exec_report:
        report_json = json.dumps(exec_report, separators=(",", ":"), ensure_ascii=False)
        user_parts.append(f"Execution report of the previous attempt:\n{report_json}")
    if feedback:
        user_parts.append(f"User feedback:\n{feedback}")

    for _attempt in range(2):
        try:
            res = client.chat.completions.create(
                model=MODEL,
                messages=[
                    {"role": "system", "content": system},
                    {"role": "user", "content": "\n\n".join(user_parts)},
                ],
                response_format={"type": "json_object"},
                max_completion_tokens=2000,
            )
            choices = getattr(res, "choices", None)
            raw = (choices[0].message.content if choices else None) or ""
        except Exception:  # noqa: BLE001 — mirrors the TS catch-all
            return None
        try:
            plan = ParsePlan.model_validate(json.loads(raw)).model_dump(by_alias=True, exclude_unset=True)
            # zod's ParsePlanSchema applies .default("keep") to onPeriodMismatch (its only
            # default); exclude_unset drops it, so back-fill to keep persisted rows byte-equal
            # to the TS stack (proposals are stored as JSON and echoed in API responses).
            for t in plan["tables"]:
                t.setdefault("onPeriodMismatch", "keep")
            validate_parse_plan(plan)
            return plan
        except Exception:  # noqa: BLE001 — JSON.parse throw, safeParse failure, and validateParsePlan throw all retry
            continue
    return None


def is_degenerate(report: dict) -> bool:
    """Should the self-check round fire? See plan Global Constraints for the pinned criteria."""
    return any(
        len(t["problems"]) > 0
        or t["rowsKept"] == 0
        or any(t["rowsKept"] > 0 and f["count"] >= t["rowsKept"] for f in t["coercionFailures"])
        for t in report["tables"]
    )
